langchain/src/agents/chat/index.ts

```diff
@@ -0,0 +1,124 @@
+import { BaseLanguageModel } from "../../schema/index.js";
+import { LLMChain } from "../../chains/index.js";
+import {
+  Agent,
+  Tool,
+  AgentInput,
+  StaticAgent,
+  staticImplements,
+  AgentStep,
+} from "../index.js";
+import {
+  HumanMessagePromptTemplate,
+  ChatPromptTemplate,
+} from "../../prompts/index.js";
+import { PREFIX, SUFFIX, FORMAT_INSTRUCTIONS } from "./prompt.js";
+
+const FINAL_ANSWER_ACTION = "Final Answer:";
+
+export type CreatePromptArgs = {
+  /** String to put after the list of tools. */
+  suffix?: string;
+  /** String to put before the list of tools. */
+  prefix?: string;
+  /** List of input variables the final prompt will expect. */
+  inputVariables?: string[];
+};
+
+type ZeroShotAgentInput = AgentInput;
+
+/**
+ * Agent for the MRKL chain.
+ * @augments Agent
+ * @augments StaticAgent
+ */
+@(staticImplements<StaticAgent>)
+export class ChatAgent extends Agent {
+  constructor(input: ZeroShotAgentInput) {
+    super(input);
+  }
+
+  _agentType() {
+    return "zero-shot-react-description" as const;
+  }
+
+  observationPrefix() {
+    return "Observation: ";
+  }
+
+  llmPrefix() {
+    return "Thought:";
+  }
+
+  _stop(): string[] {
+    return ["Observation:"];
+  }
+
+  static validateTools(tools: Tool[]) {
+    const invalidTool = tools.find((tool) => !tool.description);
+    if (invalidTool) {
+      const msg =
+        `Got a tool ${invalidTool.name} without a description.` +
+        ` This agent requires descriptions for all tools.`;
+      throw new Error(msg);
+    }
+  }
+
+  constructScratchPad(steps: AgentStep[]): string {
+    const agentScratchpad = super.constructScratchPad(steps);
+    if (agentScratchpad) {
+      return `This was your previous work (but I haven't seen any of it! I only see what you return as final answer):\n${agentScratchpad}`;
+    }
+    return agentScratchpad;
+  }
+
+  /**
+   * Create prompt in the style of the zero shot agent.
+   *
+   * @param tools - List of tools the agent will have access to, used to format the prompt.
+   * @param args - Arguments to create the prompt with.
+   * @param args.suffix - String to put after the list of tools.
+   * @param args.prefix - String to put before the list of tools.
+   */
+  static createPrompt(tools: Tool[], args?: CreatePromptArgs) {
+    const { prefix = PREFIX, suffix = SUFFIX } = args ?? {};
+    const toolStrings = tools
+      .map((tool) => `${tool.name}: ${tool.description}`)
+      .join("\n");
+    const template = [prefix, toolStrings, FORMAT_INSTRUCTIONS, suffix].join(
+      "\n\n"
+    );
+    const messages = [
+      HumanMessagePromptTemplate.fromTemplate(template),
```

Comment:
Is `SystemMessagePromptTemplate` more consistent for agent instructions?

Reply:
yes! good call